components/usage/pkg/controller/reconciler.go

```diff
@@ -38,16 +47,100 @@ func (u *UsageReconciler) Reconcile() error {
 	startOfCurrentMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
 	startOfNextMonth := startOfCurrentMonth.AddDate(0, 1, 0)
 
-	return u.reconcile(ctx, startOfCurrentMonth, startOfNextMonth)
+	status, err := u.ReconcileTimeRange(ctx, startOfCurrentMonth, startOfNextMonth)
+	if err != nil {
+		return err
+	}
+	log.WithField("usage_reconcile_status", status).Info("Reconcile completed.")
+	return nil
 }
 
-func (u *UsageReconciler) reconcile(ctx context.Context, from, to time.Time) error {
+func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.Time) (*UsageReconcileStatus, error) {
 	log.Infof("Gathering usage data from %s to %s", from, to)
-	instances, err := db.ListWorkspaceInstancesInRange(ctx, u.conn, from, to)
+	status := &UsageReconcileStatus{
+		StartTime: from,
+		EndTime:   to,
+	}
+	instances, invalidInstances, err := u.loadWorkspaceInstances(ctx, from, to)
 	if err != nil {
-		return fmt.Errorf("failed to list instances: %w", err)
+		return nil, fmt.Errorf("failed to load workspace instances: %w", err)
 	}
+	status.WorkspaceInstances = len(instances)
+	status.InvalidWorkspaceInstances = len(invalidInstances)
 
+	if len(invalidInstances) > 0 {
+		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
+	}
```

Comment:
Do we want to have metrics on how many invalid workspaces we find?

Reply:
Yep. Follow-up PRs.